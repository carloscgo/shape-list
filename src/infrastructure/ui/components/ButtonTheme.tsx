// infrastructure/ui/components/ButtonTheme.tsx

import { useCallback, useEffect, useRef } from 'react'
import { BiMoon, BiSun } from 'react-icons/bi'
import useLocalStorage from '../hooks/useLocalStorage'
import Themes from '../utils/themes'

export default function ButtonTheme() {
  const [theme, setTheme] = useLocalStorage('theme', Themes.light)

  const moon = useRef<HTMLDivElement>(null)
  const sun = useRef<HTMLDivElement>(null)

  const setDark = useCallback(
    (val: string) => {
      if (val === Themes.dark) {
        document.documentElement.classList.add(Themes.dark)
        document.documentElement.classList.remove(Themes.light)

        moon.current && moon.current.classList.add('hidden')
        sun.current && sun.current.classList.remove('hidden')
      } else {
        document.documentElement.classList.add(Themes.light)
        document.documentElement.classList.remove(Themes.dark)

        sun.current && sun.current.classList.add('hidden')
        moon.current && moon.current.classList.remove('hidden')
      }

      setTheme(val)
    },
    [moon, sun, setTheme]
  )

  useEffect(() => {
    setDark(theme)
  }, [theme, setDark])

  return (
    <>
      <div
        onClick={() => setDark(Themes.dark)}
        aria-label="moon icon"
        role="button"
        ref={moon}
        className="moon cursor-pointer text-white hover:text-blue-500 dark:hover:text-[#38BDF8]"
      >
        <BiMoon />
      </div>
      <div
        onClick={() => setDark(Themes.light)}
        aria-label="sun icon"
        role="button"
        ref={sun}
        className="sun cursor-pointer hidden text-white hover:text-blue-500 dark:hover:text-[#38BDF8]"
      >
        <BiSun />
      </div>
    </>
  )
}
