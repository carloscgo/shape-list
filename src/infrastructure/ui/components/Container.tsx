// infrastructure/ui/components/Container.tsx

import { memo } from 'react'
import ShapesNew from './Shape/New'
import ShapesList from './Shape/List'
import routes, { Route, Routes } from '../utils/routes'

const Container = memo(() => {
  return (
    <Routes>
      <Route path={routes.new} element={<ShapesNew />} />
      <Route path={routes.list} element={<ShapesList />} />
    </Routes>
  )
})

export default Container
