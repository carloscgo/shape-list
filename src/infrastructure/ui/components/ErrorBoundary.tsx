// infrastructure/ui/components/ErrorBoundary.tsx

import React, { ReactNode } from 'react'

interface ComponentProps {
  hasError?: boolean
  children: ReactNode
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Some = any

interface ComponentState {
  hasError: boolean
  error: Some
}

class ErrorBoundary extends React.Component<ComponentProps, ComponentState> {
  constructor(props: ComponentProps) {
    super(props)

    // Define a state variable to track whether is an error or not
    this.state = { hasError: props.hasError || false, error: null }
  }

  static getDerivedStateFromError(error: Some) {
    // Update state so the next render will show the fallback UI

    return { hasError: true, error }
  }

  componentDidCatch(error: Some, errorInfo: Some) {
    // You can use your own error logging service here
    console.log({ error, errorInfo })
  }

  render() {
    // Check if the error is thrown
    if (this.state.hasError) {
      // You can render any custom fallback UI
      return (
        <div>
          <h2>Oops, there is an error!</h2>
          <button
            type="button"
            onClick={() => this.setState({ hasError: false, error: null })}
          >
            Try again?
          </button>
        </div>
      )
    }

    // Some children components in case of no error
    return this.props.children
  }
}

export default ErrorBoundary
