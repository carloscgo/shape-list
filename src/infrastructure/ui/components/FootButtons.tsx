// infrastructure/ui/FootButtons.tsx

import { Link } from 'react-router-dom'
import { RiLoaderFill, RiCheckFill, RiArrowGoBackFill } from 'react-icons/ri'
import { useTranslation } from './utils/i18n'

type PropsButton = {
  routeBack: string
  label: string
  isSubmitting: boolean
  isValid: boolean
}

export default function FootButtons({
  routeBack,
  label,
  isSubmitting,
  isValid,
}: PropsButton) {
  const { t } = useTranslation()

  const isDisabled = isSubmitting || !isValid

  return (
    <div className="flex justify-center gap-4">
      <Link
        to={routeBack}
        className="my-4 capitalize bg-gray-500 text-white font-medium p-2 rounded-md hover:bg-gray-600"
      >
        <span className="flex items-center justify-center">
          <RiArrowGoBackFill size="20px" className="mr-1" />
          {t('goBack')}
        </span>
      </Link>

      <button
        disabled={isDisabled}
        type={isDisabled ? 'button' : 'submit'}
        className="my-4 capitalize bg-blue-500 text-white font-medium p-2 rounded-md hover:bg-blue-600 disabled:opacity-75 disabled:cursor-not-allowed"
      >
        {isSubmitting ? (
          <span className="flex items-center justify-center">
            <RiLoaderFill size="20px" className="animate-spin mr-1" />
            {t('creating')}
          </span>
        ) : (
          <span className="flex items-center justify-center">
            <RiCheckFill size="20px" className="mr-1" />
            {label}
          </span>
        )}
      </button>
    </div>
  )
}
