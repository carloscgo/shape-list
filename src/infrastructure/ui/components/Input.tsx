// infrastructure/ui/Input.tsx

import { ErrorMessage } from '@hookform/error-message'
import { Some } from '../../application'

type PropsInput = {
  register: (name: string, validations: object) => object
  title?: string | Some
  name: string
  type?: string
  validations: object
  errors: object
  [key: string]: number | string | object | undefined
}

export default function Input({
  register,
  title,
  name,
  type,
  validations,
  errors,
  ...props
}: PropsInput) {
  return (
    <>
      <label className="block w-100">
        {title && (
          <span className="text-gray-700 dark:text-gray-300">{title}</span>
        )}
        <input
          placeholder={title}
          {...register(name, validations)}
          name={name}
          type={type ?? 'text'}
          {...props}
          className="mt-1 mb-1 p-2 dark:bg-gray-300 block w-full rounded-md border-blue-300 border-[1px] shadow-sm focus:border-blue-300 focus:ring focus:ring-blue-200 focus:ring-opacity-50"
        />

        <ErrorMessage
          key={name}
          errors={errors}
          name={name}
          render={({ message }) => (
            <div className="bg-red-400 text-white rounded-md p-2 text-[0.8rem] w-full">
              {message}
            </div>
          )}
        />
      </label>
    </>
  )
}
