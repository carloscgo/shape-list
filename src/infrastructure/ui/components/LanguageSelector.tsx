// infrastructure/ui/LanguageSelector

import { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { fallbackLanguage, languages, key } from './utils/i18n'
import useLocalStorage from './hooks/useLocalStorage'
import styles from './styles/LanguageSelector.module.scss'

interface FlagIconProps {
  countryCode: string
}

function FlagIcon({ countryCode = '' }: FlagIconProps) {
  if (countryCode === 'en') {
    countryCode = 'gb'
  }

  return (
    <span
      className={`fi fis ${styles.fiCircle} inline-block mr-2 fi-${countryCode}`}
    />
  )
}

interface Language {
  key: string
  name: string
}

const LANGUAGE_SELECTOR_ID = 'language-selector'

export const LanguageSelector = () => {
  const [currentLanguage, setCurrentLanguage] = useLocalStorage(
    key,
    fallbackLanguage
  )

  const { i18n } = useTranslation()
  const [isOpen, setIsOpen] = useState(false)

  const selectedLanguage =
    currentLanguage ??
    languages.find((language) => language.key === i18n.language)

  const handleLanguageChange = async (language: Language) => {
    await i18n.changeLanguage(language.key)

    setCurrentLanguage(language)

    setIsOpen(false)
  }

  useEffect(() => {
    const handleWindowClick = (event: MouseEvent) => {
      const target = (event.target as HTMLElement).closest('button')
      if (target && target.id === LANGUAGE_SELECTOR_ID) {
        return
      }
      setIsOpen(false)
    }

    window.addEventListener('click', handleWindowClick)

    return () => {
      window.removeEventListener('click', handleWindowClick)
    }
  }, [])

  if (!selectedLanguage) {
    return null
  }

  return (
    <div className="flex items-center z-40">
      <div className="relative inline-block text-left">
        <div>
          <button
            onClick={() => setIsOpen(!isOpen)}
            type="button"
            className="inline-flex items-center justify-center w-full rounded-md border border-gray-300 dark:border-gray-500 shadow-sm px-4 py-2 bg-white dark:bg-gray-500 hover:shadow-md hover:shadow-sky-100 text-sm font-medium text-gray-700 dark:text-white focus:outline-none focus:ring-2 focus:ring-indigo-500"
            id={LANGUAGE_SELECTOR_ID}
            aria-haspopup="true"
            aria-expanded={isOpen}
          >
            <FlagIcon countryCode={selectedLanguage.key} />

            {selectedLanguage.name}

            <svg
              className="-mr-1 ml-2 h-5 w-5"
              xmlns="http://www.w3.org/2000/svg"
              viewBox="0 0 20 20"
              fill="currentColor"
              aria-hidden="true"
            >
              <path
                fillRule="evenodd"
                d="M10.293 14.707a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L10 12.586l3.293-3.293a1 1 0 011.414 1.414l-4 4z"
                clipRule="evenodd"
              />
            </svg>
          </button>
        </div>

        {isOpen && (
          <div
            className="origin-top-right absolute right-0 mt-2 w-full rounded-md shadow-lg bg-white dark:bg-gray-400 ring-1 ring-black ring-opacity-5"
            role="menu"
            aria-orientation="vertical"
            aria-labelledby="language-selector"
          >
            <div className="py-1 grid grid-cols-1 gap-2" role="none">
              {languages.map((language, index) => (
                <button
                  key={language.key}
                  onClick={() => handleLanguageChange(language)}
                  className={`${
                    selectedLanguage.key === language.key
                      ? 'bg-gray-100 dark:bg-gray-400 text-gray-900'
                      : 'text-gray-700 dark:bg-gray-200'
                  } px-4 py-2 text-sm text-left items-center inline-flex hover:bg-gray-100 hover:dark:bg-gray-400 ${
                    index % 2 === 0 ? 'rounded-r' : 'rounded-l'
                  }`}
                  role="menuitem"
                >
                  <FlagIcon countryCode={language.key} />
                  <span className="truncate">{language.name}</span>
                </button>
              ))}
            </div>
          </div>
        )}
      </div>
    </div>
  )
}
