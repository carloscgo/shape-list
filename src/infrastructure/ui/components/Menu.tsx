// infrastructure/ui/components/Menu.tsx

import { BiHome } from 'react-icons/bi'
import routes, { Link } from '../utils/routes'
import { useTranslation } from '../utils/i18n'

type PropsMenu = {
  maxSidebar: React.MutableRefObject<HTMLDivElement | null>
  miniSidebar: React.MutableRefObject<HTMLDivElement | null>
}

export default function Menu({ maxSidebar, miniSidebar }: PropsMenu) {
  const { t } = useTranslation()

  const ItemsMenu = [
    {
      label: t('Home'),
      url: routes.home,
      icon: <BiHome />,
    },
  ]

  return (
    <>
      <div
        ref={maxSidebar}
        className="max hidden text-white mt-20 flex-col space-y-2 w-full h-[calc(100vh)]"
      >
        {ItemsMenu.map((item, index) => (
          <Link
            key={index}
            to={item.url}
            className="hover:ml-4 w-full text-white hover:text-purple-500 dark:hover:text-blue-500 bg-[#1E293B] p-2 pl-8 rounded-full transform ease-in-out duration-300 flex flex-row items-center space-x-3"
          >
            {item.icon}
            <div>{item.label}</div>
          </Link>
        ))}
      </div>

      <div
        ref={miniSidebar}
        className="mini mt-20 flex flex-col space-y-2 w-full h-[calc(100vh)]"
      >
        {ItemsMenu.map((item, index) => (
          <Link
            key={index}
            to={item.url}
            className="hover:ml-4 justify-end pr-5 text-white hover:text-purple-500 dark:hover:text-blue-500 w-full bg-[#1E293B] p-3 rounded-full transform ease-in-out duration-300 flex"
          >
            {item.icon}
          </Link>
        ))}
      </div>
    </>
  )
}
