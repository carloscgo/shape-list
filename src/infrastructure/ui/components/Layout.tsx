// infrastructure/ui/Layout.tsx

import { useCallback, useRef } from 'react'
import { HiOutlineSquares2X2 } from 'react-icons/hi2'

import MetaTags from './MetaTags'
import Menus from './Menu'
import ButtonTheme from './ButtonTheme'
import { LanguageSelector } from './LanguageSelector'
import Container from './Container'
import styles from './styles/Layout.module.css'

const appName = import.meta.env.VITE_APP_TITLE

export default function Layout() {
  const sidebar = useRef<HTMLDivElement>(null)
  const maxSidebar = useRef<HTMLDivElement>(null)
  const miniSidebar = useRef<HTMLDivElement>(null)
  const maxToolbar = useRef<HTMLDivElement>(null)
  const logo = useRef<HTMLDivElement>(null)
  const content = useRef<HTMLDivElement>(null)

  const openNav = useCallback(() => {
    if (!sidebar.current) return

    if (sidebar.current.classList.contains('-translate-x-48')) {
      // max sidebar
      sidebar.current.classList.remove('-translate-x-48')
      sidebar.current.classList.add('translate-x-none')
      maxSidebar.current && maxSidebar.current.classList.remove('hidden')
      maxSidebar.current && maxSidebar.current.classList.add('flex')
      miniSidebar.current && miniSidebar.current.classList.remove('flex')
      miniSidebar.current && miniSidebar.current.classList.add('hidden')
      maxToolbar.current && maxToolbar.current.classList.add('translate-x-0')
      maxToolbar.current &&
        maxToolbar.current.classList.remove('translate-x-24', 'scale-x-0')
      logo.current && logo.current.classList.remove('ml-12')
      content.current && content.current.classList.remove('ml-12')
      content.current && content.current.classList.add('ml-12', 'md:ml-60')
    } else {
      // mini sidebar
      sidebar.current.classList.add('-translate-x-48')
      sidebar.current.classList.remove('translate-x-none')
      maxSidebar.current && maxSidebar.current.classList.add('hidden')
      maxSidebar.current && maxSidebar.current.classList.remove('flex')
      miniSidebar.current && miniSidebar.current.classList.add('flex')
      miniSidebar.current && miniSidebar.current.classList.remove('hidden')
      maxToolbar.current &&
        maxToolbar.current.classList.add('translate-x-24', 'scale-x-0')
      maxToolbar.current && maxToolbar.current.classList.remove('translate-x-0')
      logo.current && logo.current.classList.add('ml-12')
      content.current && content.current.classList.remove('ml-12', 'md:ml-60')
      content.current && content.current.classList.add('ml-12')
    }
  }, [sidebar, maxSidebar, miniSidebar, maxToolbar, logo, content])

  return (
    <>
      <noscript>
        <style>{`.nojs-show { opacity: 1; top: 0; }`}</style>
      </noscript>

      <MetaTags />

      <div className={`${styles.content} bg-white dark:bg-[#0F172A]`}>
        <div className="fixed w-full z-30 flex bg-white dark:bg-[#0F172A] p-2 items-center justify-center h-16 px-10">
          <div
            ref={logo}
            className="logo ml-10 dark:text-white transform ease-in-out duration-500 flex-none h-full flex items-center justify-center"
          >
            {appName}
          </div>
          <div className="grow h-full flex items-center justify-center"></div>
          <div className="flex-none h-full text-center flex items-center justify-center">
            <LanguageSelector />
          </div>
        </div>

        <aside
          ref={sidebar}
          className="w-60 -translate-x-48 fixed transition transform ease-in-out duration-1000 z-50 flex h-screen bg-[#1E293B]"
        >
          <div
            ref={maxToolbar}
            className="max-toolbar translate-x-24 scale-x-0 w-full -right-6 transition transform ease-in duration-300 flex items-center justify-between border-4 border-white dark:border-[#0F172A] bg-[#1E293B] absolute top-2 rounded-full h-12"
          >
            <div className="flex pl-4 items-center space-x-2 ">
              <div>
                <ButtonTheme />
              </div>
            </div>
            <div className="flex items-center space-x-2 group bg-gradient-to-r from-cyan-500 to-blue-500 pl-3 pr-2 py-1 rounded-full text-white">
              <div className="transform ease-in-out duration-300 mr-20">
                {appName}
              </div>
            </div>
          </div>
          <div
            onClick={() => openNav()}
            className="cursor-pointer -right-6 transition transform ease-in-out duration-500 flex border-4 border-white dark:border-[#0F172A] bg-[#1E293B] dark:hover:bg-blue-500 hover:bg-purple-500 absolute top-2 p-3 rounded-full text-white hover:rotate-45"
          >
            <HiOutlineSquares2X2 />
          </div>

          <Menus maxSidebar={maxSidebar} miniSidebar={miniSidebar} />
        </aside>

        <div
          ref={content}
          className="content ml-12 transform ease-in-out duration-500 pt-20 px-2 md:px-5 pb-4"
        >
          <Container />
        </div>
      </div>
    </>
  )
}
