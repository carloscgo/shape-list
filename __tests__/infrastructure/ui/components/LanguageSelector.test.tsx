import {
  expect,
  describe,
  it,
  render,
  fireEvent,
  React,
} from '../../../../setupTest'
import { LanguageSelector } from '../../../../src/infrastructure/ui/components/LanguageSelector'

describe('LanguageSelector', () => {
  it('should render without crashing', () => {
    const { getByLabelText } = render(<LanguageSelector />)

    const languageSelector = getByLabelText('language-selector')
    expect(languageSelector).toBeInTheDocument()
  })

  it('should show all available languages', async () => {
    const { getByRole, findByRole, getAllByText } = render(<LanguageSelector />)

    const languageSelectorButton = getByRole('button', {
      name: /English/, // change this to your default language
    })

    fireEvent.click(languageSelectorButton)

    const menu = await findByRole('menu')
    expect(menu).toBeInTheDocument()

    const languages = ['English', 'Español'] // change to your desired languages

    languages.forEach((language) => {
      const listItem = getAllByText(language)[0]
      expect(listItem).toBeInTheDocument()
    })
  })

  it('should change the language when a different language is selected', async () => {
    // setup: render component
    const { getByRole, findByRole, container } = render(<LanguageSelector />)

    // find the default language button (i.e., the currently selected language)
    const languageSelectorButton = getByRole('button', {
      name: /English/, // change this to your default language
    })

    // open the dropdown list by clicking on the button
    fireEvent.click(languageSelectorButton)

    // find the button for the target language
    const targetLanguageButton = await findByRole('menuitem', {
      name: /Español/, // change this to your target language
    })

    // simulate a click event on the target button
    fireEvent.click(targetLanguageButton)

    expect(container).toMatchSnapshot()
  })
})
