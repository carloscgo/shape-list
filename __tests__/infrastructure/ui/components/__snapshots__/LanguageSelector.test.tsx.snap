// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LanguageSelector > should change the language when a different language is selected 1`] = `
<div>
  <div
    aria-label="language-selector"
    class="flex items-center z-40"
  >
    <div
      class="relative inline-block text-left"
    >
      <div>
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="inline-flex items-center justify-center w-full rounded-md border border-gray-300 dark:border-gray-500 shadow-sm px-4 py-2 bg-white dark:bg-gray-500 hover:shadow-md hover:shadow-sky-100 text-sm font-medium text-gray-700 dark:text-white focus:outline-none focus:ring-2 focus:ring-indigo-500"
          id="language-selector"
          role="button"
          type="button"
        >
          <span
            class="fi fis _fiCircle_48a92e inline-block mr-2 fi-gb"
          />
          English
          <svg
            aria-hidden="true"
            class="-mr-1 ml-2 h-5 w-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10.293 14.707a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L10 12.586l3.293-3.293a1 1 0 011.414 1.414l-4 4z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
